Preprocessor symbols not activated during testfile syntaxtree parsing

**Describe the bug**
In my project Stryker [shows an error message](https://github.com/Testably/Testably.Abstractions/actions/runs/4902459842/jobs/8754314766), where it worked with v3.7.1:
```
[16:04:59 INF] Initial testrun started.
[16:05:15 INF] Time Elapsed 00:01:13.8443548
Unhandled exception. System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at Stryker.Core.Initialisation.ProjectMutator.EnrichTestProjectsWithTestInfo(InitialTestRun initialTestRun, TestProjectsInfo testProjectsInfo)
   at Stryker.Core.Initialisation.ProjectMutator.MutateProject(StrykerOptions options, IReporter reporters, IEnumerable`1 solutionProjects)
   at Stryker.Core.Initialisation.ProjectOrchestrator.MutateProjects(StrykerOptions options, IReporter reporters)+MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator)
   at Stryker.CLI.StrykerCli.RunStryker(IStrykerInputs inputs) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 93
   at Stryker.CLI.StrykerCli.<>c__DisplayClass10_0.<Run>b__0() in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 68
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass143_0.<OnExecute>b__0(CancellationToken _)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync(String[] args, CancellationToken cancellationToken)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Stryker.CLI.StrykerCli.Run(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 74
   at Stryker.CLI.Program.Main(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/Program.cs:line 14
/home/runner/work/_temp/e6b4cac3-c1[27](https://github.com/Testably/Testably.Abstractions/actions/runs/4902459842/jobs/8754314766#step:5:28)-4f6d-b5f5-cf92c89680a7.sh: line 2:  1858 Aborted                 (core dumped) ../../tools/dotnet-stryker -f ../.github/stryker/Stryker.Config.Testing.json -v "${GITHUB_REF#refs/heads/}" -r "Dashboard" -r "cleartext"
Error: Process completed with exit code 1[34](https://github.com/Testably/Testably.Abstractions/actions/runs/4902459842/jobs/8754314766#step:5:35).
```

**Logs**
See the [test run on Github](https://github.com/Testably/Testably.Abstractions/actions/runs/4902459842/jobs/8754314766).

**Expected behavior**
After reverting stryker to v3.7.1 the build was successful:
https://github.com/Testably/Testably.Abstractions/actions/runs/4902612340

**Desktop (please complete the following information):**
 - OS: Windows & Ubuntu
 - Type of project: framework and core
 - Framework Version N/A
 - Stryker Version 3.8.1

**Additional context**
This issue was copied from #2499

--- a/src/Stryker.Core/Stryker.Core.UnitTest/ProjectComponents/TestProjects/TestProjectTests.cs
+++ b/src/Stryker.Core/Stryker.Core.UnitTest/ProjectComponents/TestProjects/TestProjectTests.cs
@@ -1,7 +1,8 @@
 using System;
 using System.IO;
 using System.IO.Abstractions.TestingHelpers;
-using System.Reflection;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Shouldly;
 using Stryker.Core.ProjectComponents.TestProjects;
 using Xunit;
@@ -60,6 +61,31 @@ namespace Stryker.Core.UnitTest.ProjectComponents.TestProjects
 
             // Assert
             testProjectA.ShouldNotBe(testProjectB);
+        }
+
+        [Fact]
+        public void TestProject_ParseTestFile_WithCsharpParseOptions()
+        {
+            // Arrange
+            var fileSystem = new MockFileSystem();
+            var rootPath = Path.Combine("c", "TestProject");
+            var filePath = Path.Combine(rootPath, "ExampleTestFilePreprocessorSymbols.cs");
+            fileSystem.AddDirectory(rootPath);
+
+            var testProjectAnalyzerResultMock = TestHelper.SetupProjectAnalyzerResult(
+                references: Array.Empty<string>(),
+                sourceFiles: new string[] { filePath },
+                preprocessorSymbols: new string[] { "NET6_0_OR_GREATER" }
+            );
+            var file = File.ReadAllText(Path.Combine(".", "TestResources", "ExampleTestFilePreprocessorSymbols.cs"));
+            fileSystem.AddFile(filePath, new MockFileData(file));
+
+            // Act
+            var testProject = new TestProject(fileSystem, testProjectAnalyzerResultMock.Object);
+
+            // Assert
+            var nodes = testProject.TestFiles.First().SyntaxTree.GetRoot().DescendantNodes();
+            testProject.TestFiles.First().SyntaxTree.GetRoot().DescendantNodes().Where(n => n is MethodDeclarationSyntax).Count().ShouldBe(4);
         }
     }
 }

--- a/src/Stryker.Core/Stryker.Core.UnitTest/Stryker.Core.UnitTest.csproj
+++ b/src/Stryker.Core/Stryker.Core.UnitTest/Stryker.Core.UnitTest.csproj
@@ -7,6 +7,7 @@
     <Compile Remove="TestResources\ExampleSourceFile.cs" />
     <Compile Remove="TestResources\ExampleTestFileA.cs" />
     <Compile Remove="TestResources\ExampleTestFileB.cs" />
+    <Compile Remove="TestResources\ExampleTestFilePreprocessorSymbols.cs" />
   </ItemGroup>
   <ItemGroup>
     <None Remove="TestResources\StrongNameKeyFile.snk" />
@@ -15,6 +16,9 @@
     <None Remove="TestResources\FsharpExampleSourceFile.fs" />
   </ItemGroup>
   <ItemGroup>
+    <Content Include="TestResources\ExampleTestFilePreprocessorSymbols.cs">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
     <Content Include="TestResources\ExampleTestFileB.cs">
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </Content>

--- a/src/Stryker.Core/Stryker.Core.UnitTest/TestHelper.cs
+++ b/src/Stryker.Core/Stryker.Core.UnitTest/TestHelper.cs
@@ -12,7 +12,9 @@ namespace Stryker.Core.UnitTest
             IEnumerable<string> projectReferences = null,
             string targetFramework = null,
             IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>> packageReferences = null,
-            string[] references = null)
+            string[] references = null,
+            string[] preprocessorSymbols = null
+            )
         {
             var analyzerResultMock = new Mock<IAnalyzerResult>();
 
@@ -43,6 +45,10 @@ namespace Stryker.Core.UnitTest
             if (references != null)
             {
                 analyzerResultMock.Setup(x => x.References).Returns(references);
+            }
+            if (preprocessorSymbols is not null)
+            {
+                analyzerResultMock.Setup(x => x.PreprocessorSymbols).Returns(preprocessorSymbols);
             }
 
             return analyzerResultMock;

--- a/src/Stryker.Core/Stryker.Core.UnitTest/TestResources/ExampleTestFilePreprocessorSymbols.cs
+++ b/src/Stryker.Core/Stryker.Core.UnitTest/TestResources/ExampleTestFilePreprocessorSymbols.cs
@@ -1,0 +1,45 @@
+namespace ExampleProject
+{
+    public class TestClass
+    {
+
+#if NET6_0_OR_GREATER
+            
+        [Fact]
+        public int Fibinacci(int len)
+        {
+            return Fibonacci(0, 1, 1, len);
+        }
+
+        [Fact]
+        public int Fibonacci(int a, int b, int counter, int len)
+        {
+            if (counter <= len)
+            {
+                Console.Write("{0}", a);
+                return Fibonacci(b, a + b, counter + 1, len);
+            }
+            return 0;
+        }
+
+        [Fact]
+        public string LoremIpsum()
+        {
+            return @"Lorem Ipsum
+                    Dolor Sit Amet
+                    Lorem Dolor Sit";
+        }
+
+        [Fact]
+        public void StringSplit()
+        {
+            var testString = "";
+
+            testString.Split("\n");
+
+
+        }
+        
+#endif
+    }
+}

--- a/src/Stryker.Core/Stryker.Core/Initialisation/ProjectMutator.cs
+++ b/src/Stryker.Core/Stryker.Core/Initialisation/ProjectMutator.cs
@@ -60,8 +60,8 @@ namespace Stryker.Core.Initialisation
                 if (testFile is not null)
                 {
                     var lineSpan = testFile.SyntaxTree.GetText().Lines[unitTest.LineNumber - 1].Span;
-                    var node = testFile.SyntaxTree.GetRoot().DescendantNodes(lineSpan)
-                        .First(n => n is MethodDeclarationSyntax);
+                    var nodesInSpan = testFile.SyntaxTree.GetRoot().DescendantNodes(lineSpan);
+                    var node = nodesInSpan.First(n => n is MethodDeclarationSyntax);
                     testFile.AddTest(unitTest.Id, unitTest.FullyQualifiedName, node);
                 }
                 else

--- a/src/Stryker.Core/Stryker.Core/ProjectComponents/TestProjects/TestProject.cs
+++ b/src/Stryker.Core/Stryker.Core/ProjectComponents/TestProjects/TestProject.cs
@@ -32,7 +32,7 @@ namespace Stryker.Core.ProjectComponents.TestProjects
                 var syntaxTree = CSharpSyntaxTree.ParseText(sourceCode,
                     path: file,
                     encoding: Encoding.UTF32,
-                    options: new CSharpParseOptions(LanguageVersion.Latest, DocumentationMode.None));
+                    options: new CSharpParseOptions(LanguageVersion.Latest, DocumentationMode.None, preprocessorSymbols: testProjectAnalyzerResult.PreprocessorSymbols));
 
                 if (!syntaxTree.IsGenerated())
                 {

